docs/RFCS/20211106_multitenant_cluster_settings.md

```diff
@@ -0,0 +1,272 @@
+     `tenant-rw`, this value will override any setting from the tenant's side,
+     until the cluster setting is reset.
+     Note that this statement does not affect the system tenant settings.
+
+ - `ALTER TENANT <id> RESET CLUSTER SETTING <setting>`
+   - Resets the tenant setting. For `tenant-ro`, the value reverts to the `ALL`
+     value (if it is set), otherwise to the setting default. For `tenant-rw`,
+     the value reverts to the `ALL` value (if it is set), otherwise to whatever
+     value was set by the tenant (if it is set), otherwise the setting default.
+
+ - `ALTER TENANT ALL RESET CLUSTER SETTING <setting>`
+   - Resets the all-tenants setting. For tenants that have a specific value set
+     for that tenant (using `ALTER TENANT <id>`), there is no change. For other
+     tenants, `tenant-ro` values revert to the setting default and `tenant-rw`
+     values revert to whatever value was set by the tenant (if it is set),
+     otherwise the setting default.
+
+ - `SHOW CLUSTER SETTING <setting> FOR TENANT <tenant>`
+
+ - `SHOW [ALL] CLUSTER SETTINGS FOR TENANT <tenant>`
+
+New semantics for existing statements for tenants:
+ - `SHOW [ALL] CLUSTER SETTINGS` shows the `tenant-ro` and `tenant-rw` settings.
+   `Tenant-ro` settings that have an override from the host side are marked as
+   such in the description.
+
+ - `SET/RESET CLUSTER SETTING` can only be used with `tenant-rw` settings.  For
+   settings that have overrides from the host side, the statement will fail
+   explaining that the setting can only be changed once the host side resets the
+   override.
+
+## Implementation
+
+The proposed implementation is as follows:
+
+ - We update the semantics of the existing `system.settings` table:
+    - on the system tenant, this table continues to store values for all
+      settings (for the system tenant only)
+    - on other tenants, this table stores only values for `tenant-rw` settings.
+      Any table rows for other types of variables are ignored (in the case that
+      the tenant manually inserts data into the table).
+
+ - We add a new `system.tenant_settings` table with following schema:
+   ```
+   CREATE TABLE system.tenant_settings (
+     tenant_id INT8 NOT NULL,
+     name STRING NOT NULL,
+     value STRING NOT NULL,
+     value_type STRING,
+     last_updated TIMESTAMP NOT NULL DEFAULT now() ON UPDATE now(),
+     reason STRING,
+     PRIMARY KEY (tenant_id, name)
+   )
+   ```
+   This table is only used on the system tenant. All-non-system-tenant values
+   are stored in `tenant_id=0`. This table contains no settings for the system
+   tenant (`tenant_id=1`), and the `tenant_id=0` values do not apply to the
+   system tenant.
+
+ - We modify the tenant connector APIs to allow "listening" for updates to
+   cluster settings. We already have a `SystemConfigProvider` / gossip
+   subscription API with similar functionality; we can extend that or use it as
+   a precedent for a similar subsystem.
```

Comment:
I don't think we can use gossip. If by similar you mean a streaming rpc, then yes, agreed.